internal/wrphandlers/qos/qos.go

```diff
@@ -0,0 +1,138 @@
+// SPDX-FileCopyrightText: 2024 Comcast Cable Communications Management, LLC
+// SPDX-License-Identifier: Apache-2.0
+
+package qos
+
+import (
+	"context"
+	"errors"
+
+	"github.com/xmidt-org/wrp-go/v3"
+	"github.com/xmidt-org/xmidt-agent/internal/websocket"
+	"github.com/xmidt-org/xmidt-agent/internal/wrpkit"
+)
+
+var (
+	ErrInvalidInput     = errors.New("invalid input")
+	ErrMisconfiguredQOS = errors.New("misconfigured QOS")
+)
+
+// Option is a functional option type for QOS.
+type Option interface {
+	apply(*Handler) error
+}
+
+type optionFunc func(*Handler) error
+
+func (f optionFunc) apply(c *Handler) error {
+	return f(c)
+}
+
+type serviceQOSHandler func(wrp.Message) (<-chan wrp.Message, <-chan struct{})
+
+// Handler queues incoming messages and sends them to the next wrphandler
+type Handler struct {
+	// queue for wrp messages, ingested by serviceQOS
+	queue chan wrp.Message
+	// maxHeapSize is the allowable max size of the qos' priority queue, based on the sum of all queued wrp message's payload
+	maxHeapSize int
+}
+
+// New creates a new instance of the Handler struct.  The parameter next is the
+// handler that will be called and monitored for errors.
+func New(next websocket.Egress, opts ...Option) (*Handler, func(), error) {
+	if next == nil {
+		return nil, nil, ErrInvalidInput
+	}
+
+	q := make(chan wrp.Message)
+	h := &Handler{
+		queue: q,
+	}
+
+	for _, opt := range opts {
+		if opt != nil {
+			if err := opt.apply(h); err != nil {
+				return nil, nil, err
+			}
+		}
+	}
+
+	// shutdown() is used to stop serviceQOS by closing its `done` chan.
+	ctx, shutdown := context.WithCancel(context.Background())
+	go serviceQOS(ctx.Done(), h.queue, h.maxHeapSize, handleWRPWrapper(next))
+
+	return h, shutdown, nil
+}
+
+// HandleWRP queues incoming messages while the background serviceQOS goroutine attempts
+// to send as many queued messages as possible, where the highest QOS messages are prioritized
+func (h *Handler) HandleWrp(msg wrp.Message) error {
+	// h.queue will never block as long as the serviceQOS goroutine is running.
+	h.queue <- msg
+
+	return nil
+}
```

Comment:
@johnabass  should we check whether `serviceQOS`'s `done <-chan struct{}` has been closed? that way `handler.HandleWrp` can return an `io.EOF` like error when someone tries to write a message when qos's ingestion has stop?
I was thinking that'll be better than just halting because the agent's qos calls are not asynchronous, so it blocks the agent's websocket read.

Reply:
You can't check if a channel is closed.  A channel becomes signalled when it's closed, which means a `case <-done` will trigger.

If you're wanting to gate the handler, so that it returns an an error when the goroutine isn't running, a good old-fashioned mutex works.  You probably should have that anyway, to make start/stop atomic:

```
func (h *Handler) HandleWrp(msg wrp.Message) (err error) {
        h.state.Lock()
        defer h.state.Unlock()
        
	if h.queue != nil {
              // when stopped, set h.queue = nil under the same lock
	      h.queue <- msg
        } else {
              err = SomeAppropriateError
        }

	return
}
```

Reply:
I actually removed the `start` and `stop` functions to simplify qos

totally that makes sense, I was originally thinking of storing `ctx.Done()` a part of Handler to determine if qos has been shutdown.

In a situation like this, where `ctx.Done()` is basically telling us this qos is dead... create a new qos, would we want to re-use the same `ctx.Done()` that `serviceQOS` is using? I was think that would be cleaner.

```
func (h *Handler) HandleWrp(msg wrp.Message) error {
	switch {
	case <-h.done:
		return SomeAppropriateError
	default:
		// h.queue will never block as long as the serviceQOS goroutine is running.
		h.queue <- msg
	}

	return nil
}
```

Reply:
@johnabass 

I actually removed the `start` and `stop` functions to simplify qos

totally that makes sense, I was originally thinking of storing `ctx.Done()` a part of Handler to determine if qos has been shutdown.

In a situation like this, where `ctx.Done()` is basically telling us this qos is dead... create a new qos, would we want to re-use the same `ctx.Done()` that `serviceQOS` is using? I was think that would be cleaner.

```
func (h *Handler) HandleWrp(msg wrp.Message) error {
	switch {
	case <-h.done:
		return SomeAppropriateError
	default:
		// h.queue will never block as long as the serviceQOS goroutine is running.
		h.queue <- msg
	}

	return nil
}
```

Reply:
You can't reuse a context or it's done channel.  You have to create a new context and use that.  That's one reason why you need some atomic state management around starting/stopping the qos goroutine.

Reply:
I didn't know that, good to know!